### PR TITLE
[LLD][ELF] Fix SHF_MERGE misalignment when spilled

### PR DESCRIPTION
**Pull request by mysterymath**
Section merging can increase section alignment after potential spill sections are created. Since this operation is never performed on spill sections, they can keep their earlier, smaller, alignment, which produces a misalignment if a spill occurs.

This change propagates alignment increases forward after merging.

### CODE REVIEW

**Review by Prabhuk — APPROVED** [review]
Looks good to me. But I'd wait for at least one of the LLD experts to chime in.

**Review by smithp35 — APPROVED** [review]
LGTM too.

One thing that made me think is the update to the OutputSection addralign. I initially thought that we'd be at risk of increasing OutputSection alignment for the PotentialSections that aren't used, however on reflection I think we have to do that to make sure we converge.

File: lld/ELF/OutputSections.cpp [hunk 1]
```diff
@@ -253,8 +253,22 @@ void OutputSection::finalizeInputSections() {
     for (InputSection *s : isd->sections)
       commitSection(s);
   }
-  for (auto *ms : mergeSections)
+  for (auto *ms : mergeSections) {
+    // Merging may have increased the alignment of a spillable section. Update
+    // the alignment of potential spill sections and their containing output
+    // sections.
+    if (!script->potentialSpillLists.empty()) {
```
Comment (MaskRay) [inline lld/ELF/OutputSections.cpp:260]: The `empty()` test is redundant

File: lld/test/ELF/linkerscript/section-class.test [hunk 2]
```diff
@@ -310,17 +310,19 @@ SECTIONS {
 .byte 0x12, 0x34
 
 .section .b,"aM",@progbits,1
+.p2align 1
```
Comment (MaskRay) [inline lld/test/ELF/linkerscript/section-class.test:313]: An alignment of 2 works, but a larger alignment makes the effect clearer.

Comment (MaskRay) [reply lld/test/ELF/linkerscript/section-class.test:313]: We also need to dump the section content (e.g. `-x second` )

File: lld/ELF/OutputSections.cpp (hunk 1 above)
Comment (mysterymath) [reply lld/ELF/OutputSections.cpp]: Removed.

File: lld/test/ELF/linkerscript/section-class.test (hunk 2 above)
Comment (mysterymath) [reply lld/test/ELF/linkerscript/section-class.test]: Done.

**Pull request merged**